### PR TITLE
Upgrade to arrow 23.0.0

### PR DESCRIPTION
**Pull request by alamb**
# Which issue does this PR close?

re https://github.com/apache/arrow-rs/issues/2665

 # Rationale for this change
~I am putting up this PR so I can incrementally test the effect on DataFusion of splitting up the arrow crate in PRs such as  https://github.com/apache/arrow-rs/pull/2693~

~Hopefully we can use it to actually upgrade to arrow 23.0.0 when it is released as well~

Arrow 23.0.0 was released -- need to keep up with dependencies

# What changes are included in this PR?
Update to arrow 23.0.0

# Are there any user-facing changes?
No

### CODE REVIEW

**Comment by alamb** [comment]
The failure in the  `test` job https://github.com/apache/arrow-datafusion/actions/runs/3055080369/jobs/4927734152  appears to be real:

```
failures:

---- sql::expr::csv_query_nullif_divide_by_0 stdout ----
thread 'sql::expr::csv_query_nullif_divide_by_0' panicked at 'attempt to subtract with overflow', C:\Users\runneradmin\.cargo\git\checkouts\arrow-rs-3b86e19e889d5acc\5146663\arrow\src\compute\kernels\arity.rs:198:18
```

Looks related to some changes to the arity kernels  -- specifically the code that is panic'ing seems to have been changed in https://github.com/apache/arrow-rs/pull/2666 by @tustvold .  I am not sure if DataFusion is passing bad input or if there is some corner case in arrow that is hit.

I plan to look into it later today if no one else beat me to it.

**Comment by tustvold** [comment]
It will actually be https://github.com/apache/arrow-rs/pull/2643 which changed the unchecked divide kernels to panic on divide by zero - https://github.com/apache/arrow-rs/issues/2647

**Comment by alamb** [comment]
> It will actually be https://github.com/apache/arrow-rs/pull/2643 which changed the unchecked divide kernels to panic on divide by zero - https://github.com/apache/arrow-rs/issues/2647

Thanks @tustvold  -- great hint -- fixed in  https://github.com/apache/arrow-datafusion/pull/3483/commits/0493aa1855ba0529552f855da21350e7d9fb0d47

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs [hunk 1]
```diff
@@ -372,14 +372,18 @@ pub(crate) fn multiply_decimal_scalar(
...
     right: &Decimal128Array,
 ) -> Result<Decimal128Array> {
     let mul = 10_f64.powi(left.scale() as i32);
     let array = arith_decimal(left, right, |left, right| {
         let l_value = left as f64;
         let r_value = right as f64;
+        // TODO: Since this uses f64 division, divide by zero and then casting to i128
+        // gives a nasty asnwer:
+        // 170141183460469231731687303715884105727
+        //https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b5949eb324d9828a802aa11b4fa9d029
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs:386]:
```suggestion
        // casting to i128 doesn't work well when result is NaN
        // https://github.com/apache/arrow-datafusion/issues/3498
```

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 2]
```diff
@@ -24,7 +24,7 @@ use std::{any::Any, sync::Arc};
 
 use arrow::array::*;
 use arrow::compute::kernels::arithmetic::{
-    add, add_scalar, divide, divide_scalar, modulus, modulus_scalar, multiply,
+    add, add_scalar, divide_opt, divide_scalar, modulus, modulus_scalar, multiply,
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:27]: default `divide` kernel changed behavior -- `divide_opt` now has the "NULL on divide by zero" behavior datafusion expects

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs [hunk 3]
```diff
@@ -372,14 +372,18 @@ pub(crate) fn multiply_decimal_scalar(
     Ok(array)
 }
 
-pub(crate) fn divide_decimal(
+pub(crate) fn divide_opt_decimal(
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs:375]: renamed to be consistent with the arrow rename

File: datafusion/core/src/physical_plan/file_format/mod.rs [hunk 4]
```diff
@@ -286,8 +286,7 @@ impl SchemaAdapter {
         let projected_schema = Arc::new(self.table_schema.clone().project(projections)?);
 
         // Necessary to handle empty batches
-        let mut options = RecordBatchOptions::default();
-        options.row_count = Some(batch.num_rows());
+        let options = RecordBatchOptions::new().with_row_count(Some(batch.num_rows()));
```
Comment (alamb) [inline datafusion/core/src/physical_plan/file_format/mod.rs:289]: Update to use the nice API @askoa  added in https://github.com/apache/arrow-rs/pull/2729

File: dev/update_arrow_deps.py [hunk 5]
```diff
@@ -88,8 +88,20 @@ def update_version_cargo_toml(cargo_toml, new_version):
 
     for section in ("dependencies", "dev-dependencies"):
         for (dep_name, constraint) in doc.get(section, {}).items():
-            if dep_name in ("arrow", "parquet", "arrow-flight") and constraint.get("version") is not None:
-                doc[section][dep_name]["version"] = new_version
+            if dep_name in ("arrow", "parquet", "arrow-flight"):
+                if type(constraint) == tomlkit.items.String:
```
Comment (alamb) [inline dev/update_arrow_deps.py:92]: This is code I wrote in IOx in https://github.com/influxdata/influxdb_iox/blob/main/scripts/update_arrow_deps.py

Basically the previous version didn't handle `arrow = "22.0.0"` type toml

File: datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs (hunk 1 above)
Comment (alamb) [reply datafusion/physical-expr/src/expressions/binary/kernels_arrow.rs]: This has subsequently been fixed https://github.com/apache/arrow-datafusion/issues/3498 (thanks @kmitchener !)

**Review by tustvold — APPROVED** [review]

**Pull request merged**